### PR TITLE
feat(storage): add JournalReader#isEmpty() API

### LINKED ISSUE #125: Provide API for a log reader to know if the log is empty
**Is your feature request related to a problem? Please describe.**

There is currently no way for Raft log readers to know if they simply have no new elements to read or the log is completely empty, resulting in hacks to figure out if the log is empty (e.g. `reader.getFirstIndex() == 1 && reader.reset() && !reader.hasNext()`). A Journal can tell if its log is empty, so we should provide a way to distinguish the two cases.

**Describe the solution you'd like**

Ideally readers expose an `isEmpty` method which is true when there is nothing the reader can read, __regardless__ of its position in the log; meaning resetting anywhere would have no effect at all.

**Describe alternatives you've considered**

See the examples above.

**Additional context**

Related to https://github.com/zeebe-io/zeebe/issues/3543

### PR DESCRIPTION
**Pull request by npepinpe**
# Description

- provides a way for log readers to distinguish between an empty log and having no more entries to read

# Related issues

Closes #125, [zeebe-io/zeebe#3543](https://github.com/zeebe-io/zeebe/issues/3543#issuecomment-566519318)

### CODE REVIEW

**Comment by npepinpe** [comment]
NOTE: it's a little strange since typically we'll never actually use the `FileChannelJournalSegmentReader#isEmpty` but I guess that's...how it is? I also briefly considered just adding it to the `RaftLogReader` class directly and not in the interface. We could also do that; let me know what you think.

File: protocols/raft/src/main/java/io/atomix/protocols/raft/storage/log/RaftLog.java [hunk 1]
```diff
@@ -51,12 +52,12 @@ public RaftLogWriter writer() {
 
   @Override
   public RaftLogReader openReader(long index) {
-    return openReader(index, RaftLogReader.Mode.ALL);
```
Comment (ChrisKujawa) [inline protocols/raft/src/main/java/io/atomix/protocols/raft/storage/log/RaftLog.java:54]: Why we use now the different enum?

File: storage/src/test/java/io/atomix/storage/journal/AbstractJournalTest.java [hunk 2]
```diff
@@ -91,6 +92,47 @@ public static Collection primeNumbers() {
...
+    // when
+    journal.writer().append(new TestEntry(8));
+    final boolean empty = reader.isEmpty();
+
+    // then
+    assertFalse(empty);
+  }
+
+  @Test
+  public void shouldBeEmptyIfNothingCommitted() {
```
Comment (ChrisKujawa) [inline storage/src/test/java/io/atomix/storage/journal/AbstractJournalTest.java:123]: Please add an test that it is no longer empty when you commit the entry

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]
Looks good just small things

File: protocols/raft/src/main/java/io/atomix/protocols/raft/storage/log/RaftLog.java (hunk 1 above)
Comment (npepinpe) [reply protocols/raft/src/main/java/io/atomix/protocols/raft/storage/log/RaftLog.java:54]: As suggested by Sonar: https://jira.sonarsource.com/browse/RSPEC-3252

Basically `RaftLogReader` inherits from `JournalReader`, where the `Mode` enum is actually define, but it was being accessed through the child class. It's clearer to access the enum directly than indirectly, especially since we gain nothing from accessing it indirectly imo.

File: storage/src/test/java/io/atomix/storage/journal/AbstractJournalTest.java (hunk 2 above)
Comment (npepinpe) [reply storage/src/test/java/io/atomix/storage/journal/AbstractJournalTest.java:123]: Oops, I was sure I had added that one :thinking:

**Comment by npepinpe** [comment]
Updated the PR with the missing test :+1:

File: protocols/raft/src/main/java/io/atomix/protocols/raft/storage/log/RaftLog.java (hunk 1 above)
Comment (ChrisKujawa) [reply protocols/raft/src/main/java/io/atomix/protocols/raft/storage/log/RaftLog.java:54]: Ok thanks agree looks also weird I thought these both are different

**Review by ChrisKujawa — APPROVED** [review]
thanks

**Comment by npepinpe** [comment]
The `DistributedLogTest` failed but they are also flaky on `develop` (and I can reproduce it locally), so it's not related to this PR.

**Comment by npepinpe** [comment]
bors retry
